middle_end/flambda2/from_lambda/lambda_to_flambda.ml

```diff
@@ -77,7 +77,58 @@ module Env : sig
 
   val get_mutable_variable : t -> Ident.t -> Ident.t
 
-  val entering_region : t -> Ident.t -> closure_continuation:Continuation.t -> t
+  (** About local allocation regions:
+
+      In this pass, we have to transform [Lregion] expressions in Lambda to
+      primitives that mark the opening and closing of stack regions. We need to
+      ensure regions are always closed so as to not leak out of their scope.
+      They must also never be closed twice.
+
+      Several nested regions can be closed with one primitive as [End_region id]
+      which will close [id] and every other region opened in its scope. As such,
+      the transformation doesn't need to generate strict pairings of
+      [Begin_region] and [End_region] in every case. We may jump out of the
+      scope of several regions at once, in particular with exception raises from
+      [Lstaticraise].
+
+      Another case requiring attention is function calls in tail position for
+      which we may need to add an [End_region] before the jump.
+
+      This implementation works as follows.
+
+      For normal control flow, following the block structure of Lambda
+      expressions, we insert a new continuation (called the "region closure
+      continuation") upon encountering [Begin_region]; then at every leaf we
+      cause the control flow to jump via that continuation. The region closure
+      continuation closes the relevant region before jumping to what would have
+      been the "real" continuation of the leaf expressions in question. The
+      insertion of the continuation avoids duplication of the [End_region]
+      constructs. (We only need one [Begin_region] per region, but potentially
+      as many [End_region]s as there are leaves in the subsequent term.)
+
+      it suffices to jump to the closure continuation for the outermost open
+      region upon encountering a leaf in the expression tree.
+
+      For exceptional control flow, the region closure continuation is not used;
+      instead, [End_region] operations are inserted explicitly having observed
+      the difference in regions open at the time of exception raise versus at
+      the handler.
```

Comment:
This paragraph is misleading. For exceptional control flow, we open regions before the beginning of a Trywith, so that we can use this region to close an arbitrary, statically-unknown number of regions at the beginning of the handler. We certainly can't observe the difference in regions between the raise and the handler.